### PR TITLE
seccomp: add support for SECCOMP_RET_NOTIF_USER

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: lxd/container_lxc.go [hunk 1]
```diff
@@ -1811,6 +1811,13 @@ func (c *containerLXC) initLXC(config bool) error {
 		return err
 	}
 
+	if lxc.HasApiExtension("seccomp_notify") && c.DaemonState().OS.SeccompListener {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Should also be conditional so it won't run inside a truly privileged container (IsPrivileged && !inUserns)

Comment (brauner) [reply lxd/container_lxc.go]: you mean (!IsPrivileged || inUserns)?

File: lxd/container_lxc.go [hunk 2]
```diff
@@ -1811,6 +1811,13 @@ func (c *containerLXC) initLXC(config bool) error {
 		return err
 	}
 
+	if !c.isCurrentlyPrivileged() && !c.state.OS.RunningInUserNS && lxc.HasApiExtension("seccomp_notify") && c.DaemonState().OS.SeccompListener {
```
Comment (stgraber) [inline lxd/container_lxc.go]: This is done at startup, so IsPrivileged() would actually be easier to read here, isCurrentlyPrivileged is only really relevant for running containers.

File: shared/util_linux_cgo.go [hunk 3]
```diff
@@ -273,6 +287,21 @@ func AbstractUnixReceiveFd(sockFD int) (*os.File, error) {
 	return file, nil
 }
 
+func AbstractUnixReceiveFdData(sockFD int, buf []byte) (int, error) {
```
Comment (stgraber) [inline shared/util_linux_cgo.go]: Can you make adding this its own commit.

File: shared/util_linux_cgo.go [hunk 4]
```diff
@@ -208,20 +211,31 @@ int lxc_abstract_unix_recv_fds(int fd, int *recvfds, int num_recvfds,
...
+
+		memset(recvfds, -1, num_recvfds * sizeof(int));
+		if (cmsg &&
+		    cmsg->cmsg_len == CMSG_LEN(num_recvfds * sizeof(int)) &&
+		    cmsg->cmsg_level == SOL_SOCKET)
+			memcpy(recvfds, CMSG_DATA(cmsg), num_recvfds * sizeof(int));
+		break;
 	}
 
+out:
```
Comment (stgraber) [inline shared/util_linux_cgo.go]: Can you make those changes/fixes its own commit?

**Comment by stgraber** [comment]
Just a couple of nits as fixes to shared should be in a separate commit and so should the new function, will make backporting easier.

Also, you'll need an API extension for this, please add one to share/version/api.go and doc/api-extensions.md please.

**Comment by brauner** [comment]
@stgraber, updated.
